### PR TITLE
write down swarm dialing logic

### PR DESCRIPTION
**Pull request by whyrusleeping**
I thought i'd take a stab at describing the rather complicated dialing logic in a spec

### CODE REVIEW

**Comment by RichardLitt** [comment]
The code should be indented or in ```.

**Comment by diasdavid** [comment]
Was there anything added on https://github.com/ipfs/go-libp2p/pull/38 that wasn't described in these notes, @whyrusleeping ? Otherwise LGTM for merge :)

**Comment by whyrusleeping** [comment]
@diasdavid no, i think the gist of it is here

File: libp2p/8-implementations.md [hunk 1]
```diff
@@ -55,3 +55,68 @@ Current known implementations (or WIP) are:
...
+		} else {
+			rateLimitScheduleDial(peerID, addr)
+		}
+	})
+}
+
+// doDialAsync starts dialing to a specific address without blocking.
+// when the dial returns, it releases rate limit tokens, and if it
+// succeeded, will finalize the dial process.
+doDialAsync(peerID, addr) {
```
Comment (whyrusleeping) [inline libp2p/8-implementations.md]: One 'bug' i just caught in writing more tests for this is that we shouldnt start dials if they have been cancelled. For example, I queue up 100 dials, and one in the first ten or so completes successfully we don't want to continue dialing the remaining 90 queued addresses for that peer.

cc @diasdavid

Comment (diasdavid) [inline libp2p/8-implementations.md]: This finding goes along the lines of https://github.com/diasdavid/js-libp2p-swarm/issues/24

**Pull request merged**